sofiAPI.py

```diff
@@ -0,0 +1,634 @@
+
+        logger.info("Collecting account information...")
+        # Collect all account links
+        account_boxes = WebDriverWait(driver, 60).until(
+            EC.presence_of_all_elements_located((By.CSS_SELECTOR, '.AccountCardWrapper-PyEjZ .linked-card > a'))
+        )
+
+        account_dict = {}
+        for index, account_box in enumerate(account_boxes):
+            try:
+                # Open the account link in a new tab
+                account_link = account_box.get_attribute('href')
+                driver.execute_script(f"window.open('{account_link}', '_blank');")
+                driver.switch_to.window(driver.window_handles[-1])  # Switch to the newly opened tab
+
+                # Wait for the account page to load
+                WebDriverWait(driver, 60).until(check_if_page_loaded)
+
+                # Extract account number from the account page
+                account_number_element = WebDriverWait(driver, 60).until(
+                    EC.presence_of_element_located((By.CSS_SELECTOR, '#page-wrap > section:nth-child(2) > div.StyledFlex-hyCQyL.dPIcoC.HeaderWithSearchWrapper-Xeaqb.AccountHeaderWrapper-cvufWC.cSaibV.hHdLiB > div.AccountHeader-edyzFd.rVUTO > h1'))
+                )
+                account_number_text = account_number_element.text.strip()
+                account_number = account_number_text.split("#")[1].split(")")[0].strip()
+                logger.info("Account number: %s", account_number)
+
+                # Extract total value
+                current_value_element = WebDriverWait(driver, 60).until(
+                    EC.presence_of_element_located((By.CSS_SELECTOR, '#page-wrap > section:nth-child(6) > div:nth-child(3) > div > span:nth-child(2)'))
+                )
+                current_value = current_value_element.text.strip().replace('$', '').replace(',', '')
+                logger.info("Current value for account %s: %s", account_number, current_value)
+
+                # Assuming a default account type, you can adjust this if needed
+                account_type = "Investment Account"
+
+                # Store account info in the dictionary
+                account_dict[account_number] = {
+                    'type': account_type,
+                    'balance': float(current_value),
+                }
+
+                # Close the tab after extracting the information
+                driver.close()
+                driver.switch_to.window(driver.window_handles[0])
+
+            except Exception as e:
+                logger.error("Error processing account information for account %d: %s", index + 1, e)
+                # Close the tab if any error occurs and switch back
+                driver.close()
+                driver.switch_to.window(driver.window_handles[0])
+                continue
+
+        if not account_dict:
+            raise Exception("No accounts found or elements were missing.")
+
+        return account_dict
+    except TimeoutException as e:
```

Comment:
Since the actions here are the same for `TimeoutException` and `Exception`, just catch the `Exception`